File: clients/loghog.php
<?php

class Loghog {
    
    const LVL_DEBUG = 0;
    const LVL_INFO = 1;
    const LVL_WARNING = 2;
    const LVL_ERROR = 3;
    const LVL_CRITICAL = 4;

    const VERSION = 1;

    const HMAC_DIGEST_ALGO = 'md5';

    protected static $HASHABLE_FIELDS = array('app_id', 'module', 'stamp', 'nsecs', 'body');

    const USE_GZIP = 0x01;

    const _FLAGS_GZIP = 0x01;

    const FORMAT = 'NNa*';

    const DEFAULT_PORT = 5566;


    public function __construct($app_name, $options = array()) {
        $this->app_name = $app_name;

        $this->address = isset($options['address']) ? $options['address'] : 'localhost';
        $this->port = isset($options['port']) ? $options['port'] : self::DEFAULT_PORT;
        $this->stream = isset($options['stream']) ? (bool) $options['stream'] : true;
        $this->secret = isset($options['secret']) ? $options['secret'] : null;
        $this->compression = isset($options['compression']) ? $options['compression'] : null;
        $this->hostname = isset($options['hostname']) ? $options['hostname'] : php_uname('n');

        $this->pemfile = isset($options['pemfile']) ? $options['pemfile'] : null;
        $this->cacert = isset($options['cacert']) ? $options['cacert'] : null;

        $this->flags = 0;
        if ($this->compression) {
            $this->flags |= self::FLAGS_GZIP;
        }
        
        $this->sock = null;
    }

    protected function resolve($address) {
        $records = dns_get_record($this->address);
        shuffle($records);

        foreach ($records as $record) {
            if ($record['type'] == 'A' ||  $record['type'] == 'AAAA') {
                return array(
                    'ip' => $record['type'] == 'A' ? $record['ip'] : $record['ipv6'],
                    'af' => $record['type'] == 'A' ? AF_INET : AF_INET6
                );
            }
        }

        return null;
    }

    protected function make_socket($timeout=1.0) {
        if ($this->sock) {
            return;
        }

        if ($this->pemfile) {
            $opts = array(
                'ssl' => array(
                    'allow_self_signed' => false,
                    'verify_peer' => true,
                    'cafile' => $this->cacert,
                    'local_cert' => $this->pemfile
                )
            );
            $sc = stream_context_create($opts);

            $addr = sprintf('tls://%s:%d', $this->address, $this->port);
            $sock = @stream_socket_client($addr, $errno, $errstr, 1, STREAM_CLIENT_CONNECT, $sc);
        }
        else {
            if ($this->stream) {
                $addr = sprintf('tcp://%s:%d', $this->address, $this->port);
                $sock = @stream_socket_client($addr);
            }
            else {
                $addr = sprintf('udp://%s', $this->address);
                $sock = @fsockopen($addr, $this->port);
            }
        }

        if (!$sock) {
            return;
        }

        $this->sock = $sock;
    }

    private function encode($msg) {
        $time = microtime(true);
        $secs = (int) $time;
        $msecs = $time - $secs;

        $data = array(
            'version' => self::VERSION,
            'app_id' => $this->app_name,
            'module' => '', # XXX
            'stamp' => $secs,
            'nsecs' => (int) ($msecs * pow(10, 9)),
            'hostname' => $this->hostname,
            'body' => $msg 
        );

        if ($this->secret) {
            foreach (self::$HASHABLE_FIELDS as $field) {
                $hashable[$field] = (string) $data[$field];
            }
            $hashable = implode('', $hashable);
            $data['signature'] = hash_hmac(self::HMAC_DIGEST_ALGO, $hashable, $this->secret);
        }

        $payload = json_encode($data);

        if ($this->compression == self::USE_GZIP) {
            $payload = gzcompress($payload);
        }

        $size = strlen($payload);

        return pack(self::FORMAT, $size, $this->flags, $payload);
    }

    protected function emit($level, $args) {
        if ($level < 0) {
            return; # XXX
        }

        $msg = $this->format($args);
        $encoded = $this->encode($msg);

        $this->send($encoded);
    }

    protected function send($msg) {
        $this->make_socket(); // XXX: Add exponential backoff
        if (!$this->sock) {
            return;
        }

        if ($this->stream) {
            $n = @fwrite($this->sock, $msg);
        }
        else {
            $n = @fwrite($this->sock, $msg);
        }
        
        if (!$n) {
            fclose($this->sock);
            $this->sock = null;
        }
    }

    protected function format($args) {
        return vsprintf($args[0], array_slice($args, 1));
    }

    public function debug() {
        $this->emit(self::LVL_DEBUG, func_get_args());
    }

    public function info() {
        $this->emit(self::LVL_DEBUG, func_get_args());
    }

    public function warning() {
        $this->emit(self::LVL_DEBUG, func_get_args());
    }

    public function error() {
        $this->emit(self::LVL_DEBUG, func_get_args());
    }

    public function critical() {
        $this->emit(self::LVL_DEBUG, func_get_args());
    }

    public function exception() {
        $this->emit(self::LVL_ERROR, func_get_args());
    }

}

File: clients/loghog.py

import socket, hmac, hashlib, struct, zlib, ssl, random
import logging, logging.handlers

try:
    import json
except ImportError:
    import simplejson as json

class LoghogHandler(logging.handlers.SocketHandler):

    VERSION = 1
    DGRAM = 0x01
    STREAM = 0x02

    USE_GZIP = 0x01

    _FLAGS_GZIP = 0x01

    FORMAT_PROTO = '!LL %ds'

    HMAC_DIGEST_ALGO = hashlib.md5

    def __init__(self, app_name, address=('localhost', 5566), mode=STREAM, secret=None, compression=None, hostname=None, ssl_info=None):
        logging.Handler.__init__(self)

        self.app_name = app_name
        self.address = address
        self.mode = mode
        self.secret = secret
        self.compression = compression 
        self.hostname = hostname
        self.compression = None

        self.keyfile = None
        self.certfile = None
        self.cafile = None

        if ssl_info:
            self.keyfile = ssl_info['keyfile']
            self.certfile = ssl_info['certfile']
            self.cafile = ssl_info['cafile']

        if not hostname:
            self.hostname = socket.gethostname()

        self.flags = 0
        if self.compression:
            self.flags |= self._FLAGS_GZIP

        self.sock = None
        self.closeOnError = 0
        self.retryTime = None
        #
        # Exponential backoff parameters.
        #
        self.retryStart = 1.0
        self.retryMax = 30.0
        self.retryFactor = 2.0

    def _resolve_addr(self, address, mode):
        '''Resolves the given address and mode into a randomized address record.'''

        socktype = socket.SOCK_DGRAM if mode == self.DGRAM else socket.SOCK_STREAM
        res = socket.getaddrinfo(address[0], address[1], 0, socktype)
        random.shuffle(res)
        return res[0]

    def makeSocket(self, timeout=1.0):
        '''Makes a connection to the socket.'''

        af, socktype, proto, cannonname, sa = self._resolve_addr(self.address, self.mode)
        
        s = socket.socket(af, socktype, proto)

        if hasattr(s, 'settimeout'):
            s.settimeout(timeout)

        if self.mode == self.STREAM:
            if self.keyfile:
                s = ssl.wrap_socket(s,
                    keyfile=self.keyfile,
                    certfile=self.certfile,
                    server_side=False,
                    cert_reqs=ssl.CERT_NONE, # XXX should be ssl.CERT_REQUIRED
                    ca_certs=self.cafile,
                    ssl_version=ssl.PROTOCOL_TLSv1,
                    do_handshake_on_connect=True,
                    suppress_ragged_eofs=True,
                    ciphers=None
                )

            try:
                s.connect(sa)
            except Exception:
                pass

        return s

    def _encode(self, record):
        '''Encodes a log record into the loghog on-wire protocol.'''

        data = {
            'version': self.VERSION,
            'app_id': self.app_name,
            'module': record.name,
            'stamp': int(record.created),
            'nsecs': int(record.msecs * 10**6),
            'hostname': self.hostname,
            'body': self.format(record),
        }

        if self.secret:
            hashable_fields = ['app_id', 'module', 'stamp', 'nsecs', 'body']
            hashable = u''.join(unicode(data[field]) for field in hashable_fields).encode('utf-8')
            data['signature'] = hmac.new(self.secret, hashable, self.HMAC_DIGEST_ALGO).hexdigest()

        payload = json.dumps(data)

        if self.compression == self.USE_GZIP:
            payload = zlib.compress(payload)

        size = len(payload)
        return struct.pack(self.FORMAT_PROTO % size, size, self.flags, payload)

    def emit(self, record):
        '''Encodes and sends the messge over the network.'''

        self.send(self._encode(record))

    def send(self, s):
        '''Attempts to create a network connection and send the data.'''

        if self.sock is None:
            self.createSocket()

        if not self.sock:
            return

        try:
            if self.mode == self.DGRAM:
                self.sock.sendto(s, self.address)
            elif self.mode == self.STREAM:
                self.sock.sendall(s)
        except socket.error:
            self.close()

